projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+const (
+	url        = "http://localhost:8080"
+	maxRetries = 3
+)
+
+func main() {
+	currentTime := time.Now()
+	client := &http.Client{}
+
+	for retries := 0; retries <= maxRetries; retries++ {
+		response, err := handleWeatherRequest(url, client, currentTime, retries)
+		if err != nil {
+			handleError(err)
```

Comment:
Now that you only call this in one place, I'd probably inline it here.

Reply:
I did following you comment about `defer` , did I misunderstood it ?
FYI defers get called at the end of the function, whereas I suspect you really want this to happen at the end of the loop iteration. In this instance it doesn't really matter (the program will be done soon enough regardless), but it's good to know e.g. if you're writing a server that loops for a long time you'll end up leaking memory if you defer inside a loop.

To fix this, you could pull out a function which does "the contents of the for loop" :)

Reply:
Yeah, for the rest of the behaviour I think this looks great, but `handleError` doesn't have any `defer`s in it, so it doesn't need to be pulled out into a separate function - it's fine if its code just sits inside the loop.